### PR TITLE
win: Add new flag `UV_INHERIT_HANDLE`

### PR DESCRIPTION
**Pull request by moteus**
Allows set HANDLE value as stdio handles for `uv_spawn` function.
On not Windows systems this flag has same effect as UV_INHERIT_FD.

### CODE REVIEW

File: src/unix/process.c [hunk 1]
```diff
@@ -180,6 +180,8 @@ int uv__make_pipe(int fds[2], int flags) {
   return 0;
 }
 
+/* Verify that HANDLE is same as int*/
+STATIC_ASSERT(sizeof(((uv_stdio_container_t*)0)->data.handle) == sizeof(((uv_stdio_container_t*)0)->data.fd));
```
Comment (bnoordhuis) [inline src/unix/process.c]: Can you keep lines <= 80 columns?

Also, in the comment above, can you end with `...is same as int. */`, i.e., punctuation and a blank?

File: src/unix/process.c [hunk 2]
```diff
@@ -189,7 +191,7 @@ static int uv__process_init_stdio(uv_stdio_container_t* container, int fds[2]) {
   int mask;
   int fd;
 
-  mask = UV_IGNORE | UV_CREATE_PIPE | UV_INHERIT_FD | UV_INHERIT_STREAM;
+  mask = UV_IGNORE | UV_CREATE_PIPE | UV_INHERIT_FD | UV_INHERIT_STREAM | UV_INHERIT_HANDLE;
```
Comment (bnoordhuis) [inline src/unix/process.c]: Long line.  I'll stop pointing that out from now on.

File: src/win/process-stdio.c [hunk 3]
```diff
@@ -347,21 +349,36 @@ int uv__stdio_create(uv_loop_t* loop,
...
+        }
+        else {
+          HANDLE base_handle = fdopt.data.handle;
+          if (base_handle == INVALID_HANDLE_VALUE) {
             CHILD_STDIO_CRT_FLAGS(buffer, i) = 0;
             CHILD_STDIO_HANDLE(buffer, i) = INVALID_HANDLE_VALUE;
             break;
           }
-          goto error;
+          /*Duplicate handle because finalizer close this handle*/
```
Comment (bnoordhuis) [inline src/win/process-stdio.c]: Blanks and punctuation here, please.

File: src/win/process-stdio.c [hunk 4]
```diff
@@ -260,6 +260,9 @@ int uv__create_nul_handle(HANDLE* handle_ptr,
   return 0;
 }
 
+/* Verify that HANDLE can be stored in `stream` field and not increase struct size. */
```
Comment (bnoordhuis) [inline src/win/process-stdio.c]: This comment still goes over the 80 column limit.

**Comment by cjihrig** [comment]
@moteus can you add some tests, per https://github.com/libuv/libuv/pull/1121#pullrequestreview-7060581.

**Comment by moteus** [comment]
I think for test can be used `spawn_inherit_streams` with few changes.

**Comment by moteus** [comment]
Ping

**Comment by saghul** [comment]
The name of the flag is a bit misleading, as one could think libuv handles can be inherited.  I suggest adding "WINDOWS" to the flag name.

Docs are also missing.

@moteus what is the motivation for this?   We are kinda planning to get away from fds on Windows (see https://github.com/libuv/leps/pull/8) so TBH I'd rather have this patch change the `fd` field to type `uv_os_fd_t` and maybe rename the flag.  That means it would have to go to master though.

**Comment by moteus** [comment]
>The name of the flag is a bit misleading, as one could think libuv handles can be inherited. I suggest adding "WINDOWS" to the flag name.

Like `UV_INHERIT_WINDOWS_HANDLE`? But what about other platforms?
Now `UV_INHERIT_HANDLE` has same meaning on *nix platforms like `UV_INHERIT_FD`.
So you can use it in crossplatform code and just use FD on non Windows platforms.
May be `UV_INHERIT_SYSTEM_HANDLE` or `UV_INHERIT_RAW_HANDLE`?

> what is the motivation for this?

My usecase is use `uv_spawn` to run PHP in fcgi mode. And for this you need create socet bind/listen it and pass this socket to php child processes as stdin handle. Now I convert socket to fd and pass it to libuv. But It require that both app and libuv (I use dynamic link) use same runtime. And it will not works for static runtime.

> We are kinda planning to get away from fds on Windows

I think using fd has its own usecases. I think it convinient to pass stdio as fd 0,1,2 instead of call GetStdHandle function to obtain it. E.g. I use libuv in Lua and there no existed binding to this function.

**Comment by saghul** [comment]
> Like UV_INHERIT_WINDOWS_HANDLE? But what about other platforms?
> Now UV_INHERIT_HANDLE has same meaning on *nix platforms like UV_INHERIT_FD.
> So you can use it in crossplatform code and just use FD on non Windows platforms.
> May be UV_INHERIT_SYSTEM_HANDLE or UV_INHERIT_RAW_HANDLE?

There you go.  It means roughly the same thing, but with a different name per platform.  I'd like to have a single flag which does the right thing on each, fds on Unix, handles on Windows, moving forward.

> I think using fd has its own usecases. I think it convinient to pass stdio as fd 0,1,2 instead of call GetStdHandle function to obtain it. E.g. I use libuv in Lua and there no existed binding to this function.

Please read the LEP, it's not that simple.

Overall I'm a bit concerned about this patch not being future-proof.

**Comment by moteus** [comment]
I agree it possible just use only HANDLES and not fd on Windows. 
Like I say I have nothing against this. It just breaks ABI. Also it not urgent
feature for me because I have workaround for current LibUV version.

**Comment by saghul** [comment]
@moteus Great! Can you re-target this at master then?  Feel free to break the ABI there :-)

**Comment by moteus** [comment]
If you have plans to implement libuv/leps#8 then I see no point in my PR.

**Comment by saghul** [comment]
Right, so are ok with closing this and revisiting when the LEP implementation is in place?

**Comment by moteus** [comment]
Yes. But I think #1119 should be opened until it was implemented. May be can be marked for next major versoin.

**Comment by saghul** [comment]
Ack. I added the v2 label there. Thanks!

**Pull request closed without merging**